src/bundle.c

```diff
@@ -517,159 +555,155 @@ enum swupd_code remove_bundles(char **bundles)
-			struct list *iter;
-			char *bundle;
-			iter = list_head(reqd_by);
-			error("bundle requested to be removed is required by the following bundles:\n");
-			info("format:\n");
-			info(" # * is-required-by\n");
-			info(" #   |-- is-required-by\n");
-			info(" # * is-also-required-by\n # ...\n\n");
-			while (iter) {
-				bundle = iter->data;
-				iter = iter->next;
-				info("%s", bundle);
+				/* the --force option was specified */
+				info("\nThe --force option was used, bundle \"%s\" and all bundles that require it will be removed from the system\n", bundle);
+
+				/* move the manifest of dependent bundles to the list of bundles to be removed */
+				struct list *iter = NULL;
+				char *dep;
+				iter = list_head(reqd_by);
+				while (iter) {
+					dep = iter->data;
+					iter = iter->next;
+					list_move_item(dep, &current_mom->submanifests, &bundles_to_remove, find_manifest);
+					remove_tracked(dep);
+				}
+				list_free_list_and_data(reqd_by, free);
+				reqd_by = NULL;
 			}
-
-			list_free_list_and_data(reqd_by, free);
-			ret = SWUPD_REQUIRED_BUNDLE_ERROR;
-			bad++;
-			goto out_free_mom;
 		}
 
-		current_mom->files = consolidate_files_from_bundles(current_mom->submanifests);
+		/* move the manifest of the bundle to be removed from the list of subscribed bundles
+		 * to the list of bundles to be removed */
+		list_move_item(bundle, &current_mom->submanifests, &bundles_to_remove, find_manifest);
+		info("\nRemoving bundle: %s\n", bundle);
+		remove_tracked(bundle);
+	}
 
-		/* Now that we have the consolidated list of all files, load bundle to be removed submanifest */
-		ret = load_bundle_manifest(bundle, subs, current_version, &bundle_manifest);
-		if (ret != 0 || !bundle_manifest) {
-			error("Cannot load %s sub-manifest (ret = %d)\n", bundle, ret);
-			bad++;
-			goto out_free_mom;
-		}
+	/* get the list of all files required by the installed bundles (except the ones to be removed) */
+	current_mom->files = consolidate_files_from_bundles(current_mom->submanifests);
 
-		/* deduplication needs file list sorted by filename, do so */
-		bundle_manifest->files = list_sort(bundle_manifest->files, file_sort_filename);
-		deduplicate_files_from_manifest(&bundle_manifest, current_mom);
-
-		info("Deleting bundle files...\n");
-		remove_files_in_manifest_from_fs(bundle_manifest);
-		remove_tracked(bundle_manifest->component);
-
-		free_manifest(bundle_manifest);
-	out_free_mom:
-		free_manifest(current_mom);
-	out_free_curl:
-		telemetry(ret ? TELEMETRY_CRIT : TELEMETRY_INFO,
-			  "bundleremove",
-			  "bundle=%s\n"
-			  "current_version=%d\n"
-			  "result=%d\n"
-			  "bytes=%ld\n",
-			  bundle,
-			  current_version,
-			  ret,
-			  total_curl_sz);
-		/* if at least one of the bundles fails to be removed, exit with a failure */
-		if (ret) {
-			ret_code = ret;
-		}
+	/* get the list of the files contained in the bundles to be removed */
+	files_to_remove = consolidate_files_from_bundles(bundles_to_remove);
```

Comment:
I found the performance problem I was complaining about. It's not a big deal, but we don't need to calculate that if there's no bundle to be installed.

Reply:
Fixed it